### ISSUE TITLE
twistd --help plugin list isn't alphabetized

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3077]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#3077|
|Type|enhancement|
|Created|2008-03-16 02:26:47Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-help-sort-3077-2|

It really should be.

Attachments:

* [twistd-help-sort-3077.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/58d/58d2c15b44da0ce544f22ca5763b73f6e823ca79/86f7a067d2deab11ea6eedf204b1cfff265c39aa.patch) (6077 bytes) - added by indigo on 2011-10-04 14:53:01Z - with unit tests

<details><summary>Searchable metadata</summary>

```
trac-id__3077 3077
type__enhancement enhancement
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__branches_twistd_help_sort_3077_2 branches/twistd-help-sort-3077-2
branch_author__thijs__indigo thijs, indigo
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation_twistd documentation twistd
time__1205634407000000 1205634407000000
changetime__1317759419000000 1317759419000000
version__None None
owner__exarkun exarkun
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [30505]) Branching to 'twistd-help-sort-3077'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [30507]) sort twistd subcommands and reactors list in help. refs [#3077](https://github.com/twisted/twisted/issues/3077)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

`twistd --help` now returns:

```
Commands:
    conch            A Conch SSH service.
    dns              A domain name server.
    ftp              An FTP server.
    inetd            An inetd(8) replacement.
    mail             An email service
    manhole          An interactive remote debugger service accessible via
                     telnet and ssh and providing syntax coloring and basic line
                     editing functionality.
    manhole-old      An interactive remote debugger service.
    news             A news server.
    portforward      A simple port-forwarder.
    procmon          A process watchdog / supervisor
    socks            A SOCKSv4 proxy service.
    telnet           A simple, telnet-based remote debugging service.
    toc              An AIM TOC service.
    web              A general-purpose web server which can serve from a
                     filesystem or application resource.
    web2             An HTTP/1.1 web server that can serve from a filesystem or
                     application resource.
    words            A modern words server
    xmpp-router      An XMPP Router server
```

and `twistd --help-reactors`:

```
    cf  	CoreFoundation integration reactor.
    debug-gui	Semi-functional debugging/introspection reactor.
    default	The best reactor for the current platform.
    epoll	epoll(4)-based reactor.
    glib2	GLib2 event-loop integration reactor.
    gtk 	Gtk1 integration reactor.
    gtk2	Gtk2 integration reactor.
    iocp	Win32 IO Completion Ports-based reactor.
    kqueue	kqueue(2)-based reactor.
    poll	poll(2)-based reactor.
    qt  	QT integration reactor
    select	select(2)-based reactor.
    win32	Win32 WaitForMultipleObjects-based reactor.
    wx  	wxPython integration reactor.
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Unit tests?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_9"></a>|@bitglue commented|
|-|-|

patch with tests attached. I'm thinking this would be merged with a .misc file? Not sure the convention; I'll leave it to the reviewer.

BONUS POINTS FOR REVIEWER: change izip to zip in twisted.test.test_twistd.ServerOptionsTest.test_subCommands and re-run tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32741]) Branching to 'twistd-help-sort-3077-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [32742]) Apply twistd-help-sort-3077.patch

refs [#3077](https://github.com/twisted/twisted/issues/3077)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  a .misc file for this sounds fine to me.  I could also imagine a .feature file.  The criteria are highly subjective, of course, with the main one being a guess about what users will be most interested in.

  1. The zip/izip issue looks like a semi-important one.  You probably noticed that the loop implementing `subCommands` is not early binding the value of the `plug` loop variable.  So sorting after consuming the whole generator leads to different results than sorting while consuming the generator incrementally (presumably this is due to differences in the way the builtin `zip` and `izip` are implemented, but I didn't investigate very deeply).  Early binding the loop variable in the lambda is probably a good fix.  The lambda perhaps also merits a comment - I think it is done that way to avoid doing extra work when the options object is provided using a property, and the property getter is non-trivial.
  1. Aim for lines of at most 80 columns
  1. Stick to top-level imports except where there is a functional difference given by nesting them.

Very nice otherwise, thanks a lot for working on this.  The review points are not very deep, so I'll address them and merge if buildbot looks happy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Oh, I guess one other minor point is that I'd avoid try to avoid the extra underscores in `test_opt_help_reactors`.  It's a close thing, since I like having names reflect what they're testing, but underscores are reserved for dynamic dispatch in the coding standard, and while `opt_help_reactors` is doing that, trial stops at the `test_` when looking at test methods.  So something like `test_sortedReactorHelp` seems better to me (so I'll rename that).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [32743]) Address simple coding standard-type review feedback

refs [#3077](https://github.com/twisted/twisted/issues/3077)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32744]) Early bind the loop variable, comment about where, and trivially simplify the unit test which was working around this bug

refs [#3077](https://github.com/twisted/twisted/issues/3077)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [32746]) Merge twistd-help-sort-3077-2

Author: thijs, indigo
Reviewer: exarkun
Fixes: [#3077](https://github.com/twisted/twisted/issues/3077)

Sort the list of reactors printed by `twistd --help-reactors` and the
list of plugins printed by `twistd --help`.
